File: graph_mst/prims/src/PrimsDemo.swift
// Swift
// グラフの最小全域木: プリム法 (Prim)

// 優先度キューの代わりにヒープを使用するための実装
struct PriorityQueue<Element> {
    private var elements: [Element]
    private let hasHigherPriority: (Element, Element) -> Bool
    
    init(hasHigherPriority: @escaping (Element, Element) -> Bool) {
        self.elements = []
        self.hasHigherPriority = hasHigherPriority
    }
    
    var isEmpty: Bool {
        return elements.isEmpty
    }
    
    mutating func enqueue(_ element: Element) {
        elements.append(element)
        siftUp(from: elements.count - 1)
    }
    
    mutating func dequeue() -> Element? {
        guard !elements.isEmpty else { return nil }
        
        elements.swapAt(0, elements.count - 1)
        let element = elements.removeLast()
        if !elements.isEmpty {
            siftDown(from: 0)
        }
        return element
    }
    
    private mutating func siftUp(from index: Int) {
        var childIndex = index
        let child = elements[childIndex]
        var parentIndex = (childIndex - 1) / 2
        
        while childIndex > 0 && hasHigherPriority(child, elements[parentIndex]) {
            elements[childIndex] = elements[parentIndex]
            childIndex = parentIndex
            parentIndex = (childIndex - 1) / 2
        }
        
        elements[childIndex] = child
    }
    
    private mutating func siftDown(from index: Int) {
        let count = elements.count
        let parent = elements[index]
        var parentIndex = index
        
        while true {
            let leftChildIndex = 2 * parentIndex + 1
            let rightChildIndex = leftChildIndex + 1
            
            var bestChildIndex = parentIndex
            
            if leftChildIndex < count && hasHigherPriority(elements[leftChildIndex], elements[bestChildIndex]) {
                bestChildIndex = leftChildIndex
            }
            
            if rightChildIndex < count && hasHigherPriority(elements[rightChildIndex], elements[bestChildIndex]) {
                bestChildIndex = rightChildIndex
            }
            
            if bestChildIndex == parentIndex {
                break
            }
            
            elements[parentIndex] = elements[bestChildIndex]
            parentIndex = bestChildIndex
        }
        
        elements[parentIndex] = parent
    }
}

class GraphData {
    // 隣接リストとしてグラフデータを格納します。
    // キーは頂点、値はその頂点に隣接する頂点とその辺の重みのタプルのリストです。
    private var data: [String: [(String, Int)]] = [:]
    
    func get() -> [String: [(String, Int)]] {
        // グラフの内部データを取得します。
        return data
    }
    
    func getVertices() -> [String] {
        // グラフの全頂点をリストとして返します。
        return Array(data.keys)
    }
    
    func getEdges() -> [(String, String, Int)] {
        // グラフの全辺をリストとして返します。
        // 無向グラフの場合、(u, v, weight) の形式で返します。
        // 重複を避けるためにセットを使用します。
        var edges = Set<String>()
        var result: [(String, String, Int)] = []
        
        for (vertex, neighbors) in data {
            for (neighbor, weight) in neighbors {
                // 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
                let edge = [vertex, neighbor].sorted().joined()
                if !edges.contains(edge) {
                    edges.insert(edge)
                    result.append(([vertex, neighbor].sorted()[0], [vertex, neighbor].sorted()[1], weight))
                }
            }
        }
        
        return result
    }
    
    func getNeighbors(vertex: String) -> [(String, Int)]? {
        // 指定された頂点の隣接ノードと辺の重みのリストを返します。
        // 形式: [(隣接頂点, 重み), ...]
        return data[vertex]
    }
    
    func getEdgeWeight(vertex1: String, vertex2: String) -> Int? {
        // 指定された2つの頂点間の辺の重みを返します。
        // 辺が存在しない場合はnilを返します。
        if let neighbors = data[vertex1] {
            for (neighbor, weight) in neighbors {
                if neighbor == vertex2 {
                    return weight
                }
            }
        }
        return nil
    }
    
    @discardableResult
    func addVertex(vertex: String) -> Bool {
        // 新しい頂点をグラフに追加します。
        if data[vertex] == nil {
            data[vertex] = []
        }
        // 既に存在する場合は追加しないがTrueを返す（変更なしでも成功とみなす）
        return true
    }
    
    @discardableResult
    func addEdge(vertex1: String, vertex2: String, weight: Int) -> Bool {
        // 両頂点間に辺を追加します。重みを指定します。
        // 頂点がグラフに存在しない場合は追加します。
        if data[vertex1] == nil {
            addVertex(vertex: vertex1)
        }
        if data[vertex2] == nil {
            addVertex(vertex: vertex2)
        }
        
        // vertex1 -> vertex2 の辺を追加（重み付き）
        var edgeExistsV1V2 = false
        if var neighbors = data[vertex1] {
            for i in 0..<neighbors.count {
                if neighbors[i].0 == vertex2 {
                    neighbors[i] = (vertex2, weight) // 既に存在する場合は重みを更新
                    data[vertex1] = neighbors
                    edgeExistsV1V2 = true
                    break
                }
            }
            if !edgeExistsV1V2 {
                data[vertex1]?.append((vertex2, weight))
            }
        }
        
        // vertex2 -> vertex1 の辺を追加（重み付き）
        var edgeExistsV2V1 = false
        if var neighbors = data[vertex2] {
            for i in 0..<neighbors.count {
                if neighbors[i].0 == vertex1 {
                    neighbors[i] = (vertex1, weight) // 既に存在する場合は重みを更新
                    data[vertex2] = neighbors
                    edgeExistsV2V1 = true
                    break
                }
            }
            if !edgeExistsV2V1 {
                data[vertex2]?.append((vertex1, weight))
            }
        }
        
        return true
    }    
    
    func isEmpty() -> Bool {
        // グラフが空かどうかを返します。
        return data.isEmpty
    }
        
    @discardableResult
    func clear() -> Bool {
        // グラフを空にします。
        data = [:]
        return true
    }
    
    func getMST(startVertex: String? = nil) -> [(String, String, Int)]? {
        let vertices = getVertices()
        if vertices.isEmpty {
            return [] // グラフが空
        }
        
        let start = startVertex ?? vertices[0]
        guard data[start] != nil else {
            print("ERROR: 開始頂点 \(start) はグラフに存在しません。")
            return nil
        }
        
        // MSTに含まれる頂点のセット
        var inMST = Set<String>()
        // 優先度付きキュー (重み, 現在の頂点, 遷移元の頂点)
        var minHeap = PriorityQueue<(Int, String, String?)> { $0.0 < $1.0 }
        // MSTを構成する辺のリスト
        var mstEdges: [(String, String, Int)] = []
        // 各頂点への最小コスト（MSTに追加する際の辺の重み）と、その遷移元の頂点を記録
        var minCost = [String: Int]()
        var parent = [String: String?]()
        
        for vertex in vertices {
            minCost[vertex] = Int.max
            parent[vertex] = nil
        }
        
        // 開始頂点の処理
        minCost[start] = 0
        minHeap.enqueue((0, start, nil)) // (コスト, 現在の頂点, 遷移元の頂点)
        
        while !minHeap.isEmpty {
            // 最小コストの辺を持つ頂点を取り出す
            guard let (cost, currentVertex, fromVertex) = minHeap.dequeue() else { break }
            
            // 既にMSTに含まれている頂点であればスキップ
            if inMST.contains(currentVertex) {
                continue
            }
            
            // 現在の頂点をMSTに追加
            inMST.insert(currentVertex)
            
            // MSTに追加された辺を記録 (開始頂点以外)
            if let from = fromVertex {
                // from_vertex から current_vertex への辺の重みを取得
                if let weight = getEdgeWeight(vertex1: from, vertex2: currentVertex) {
                    let edge = ([from, currentVertex].sorted(), weight)
                    mstEdges.append((edge.0[0], edge.0[1], edge.1))
                }
            }
            
            // 現在の頂点に隣接する頂点を調べ、MSTへの追加コストを更新
            if let neighborsWithWeight = getNeighbors(vertex: currentVertex) {
                for (neighbor, weight) in neighborsWithWeight {
                    // 隣接頂点がまだMSTに含まれておらず、現在のコストよりも小さい場合
                    if !inMST.contains(neighbor) && weight < minCost[neighbor]! {
                        minCost[neighbor] = weight
                        parent[neighbor] = currentVertex
                        minHeap.enqueue((weight, neighbor, currentVertex))
                    }
                }
            }
        }
        
        return mstEdges
    }
}

func main() {
    print("Prims TEST -----> start")
    let graphData = GraphData()
    
    graphData.clear()
    let inputList1 = [("A", "B", 4), ("B", "C", 3), ("B", "D", 2), ("D", "A", 1), ("A", "C", 2), ("B", "D", 2)]
    for input in inputList1 {
        graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
    }
    print("\nグラフの頂点: \(graphData.getVertices())")
    print("グラフの辺 (重み付き): \(graphData.getEdges())")
    if let outputMst = graphData.getMST() {
        for edge in outputMst {
            print("Edge: \(edge.0) - \(edge.1), Weight: \(edge.2)")
        }
        let totalWeight = outputMst.reduce(0) { $0 + $1.2 }
        print("最小全域木の合計重み: \(totalWeight)")
    }
    
    graphData.clear()
    let inputList2 = [("A", "B", 4), ("C", "D", 4), ("E", "F", 1), ("F", "G", 1)]
    for input in inputList2 {
        graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
    }
    print("\nグラフの頂点: \(graphData.getVertices())")
    print("グラフの辺 (重み付き): \(graphData.getEdges())")
    if let outputMst = graphData.getMST() {
        for edge in outputMst {
            print("Edge: \(edge.0) - \(edge.1), Weight: \(edge.2)")
        }
        let totalWeight = outputMst.reduce(0) { $0 + $1.2 }
        print("最小全域木の合計重み: \(totalWeight)")
    }
    
    graphData.clear()
    let inputList3 = [("A", "B", 4), ("B", "C", 3), ("D", "E", 5)]
    for input in inputList3 {
        graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
    }
    print("\nグラフの頂点: \(graphData.getVertices())")
    print("グラフの辺 (重み付き): \(graphData.getEdges())")
    if let outputMst = graphData.getMST() {
        for edge in outputMst {
            print("Edge: \(edge.0) - \(edge.1), Weight: \(edge.2)")
        }
        let totalWeight = outputMst.reduce(0) { $0 + $1.2 }
        print("最小全域木の合計重み: \(totalWeight)")
    }
    
    graphData.clear()
    let inputList4: [(String, String, Int)] = []
    for input in inputList4 {
        graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
    }
    print("\nグラフの頂点: \(graphData.getVertices())")
    print("グラフの辺 (重み付き): \(graphData.getEdges())")
    if let outputMst = graphData.getMST() {
        for edge in outputMst {
            print("Edge: \(edge.0) - \(edge.1), Weight: \(edge.2)")
        }
        let totalWeight = outputMst.reduce(0) { $0 + $1.2 }
        print("最小全域木の合計重み: \(totalWeight)")
    }
    
    print("\nPrims TEST <----- end")
}

main()

File: graph_mst/prims/src/PrimsDemo.go
// Go
// グラフの最小全域木: プリム法 (Prim)

package main

import (
	"container/heap"
	"fmt"
)

// Edge は辺を表す構造体
type Edge struct {
	vertex string
	weight int
}

// MinHeapItem はヒープ内の項目を表す構造体
type MinHeapItem struct {
	cost     int
	vertex   string
	fromEdge string
	index    int
}

// MinHeap は優先度付きキューの実装
type MinHeap []*MinHeapItem

// Len はヒープの長さを返す
func (h MinHeap) Len() int { return len(h) }

// Less は小さい方が優先されるように比較する
func (h MinHeap) Less(i, j int) bool { return h[i].cost < h[j].cost }

// Swap はヒープ内の要素を交換する
func (h MinHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
	h[i].index = i
	h[j].index = j
}

// Push はヒープに要素を追加する
func (h *MinHeap) Push(x interface{}) {
	n := len(*h)
	item := x.(*MinHeapItem)
	item.index = n
	*h = append(*h, item)
}

// Pop はヒープから最小要素を取り出す
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	item := old[n-1]
	old[n-1] = nil  // メモリリーク防止
	item.index = -1 // 安全のため
	*h = old[0 : n-1]
	return item
}

// GraphData はグラフを表す構造体
type GraphData struct {
	data map[string][]Edge
}

// NewGraphData は新しいGraphDataインスタンスを作成する
func NewGraphData() *GraphData {
	return &GraphData{
		data: make(map[string][]Edge),
	}
}

// Get はグラフの内部データを取得する
func (g *GraphData) Get() map[string][]Edge {
	return g.data
}

// GetVertices はグラフの全頂点をスライスとして返す
func (g *GraphData) GetVertices() []string {
	vertices := make([]string, 0, len(g.data))
	for v := range g.data {
		vertices = append(vertices, v)
	}
	return vertices
}

// GetEdges はグラフの全辺をスライスとして返す
func (g *GraphData) GetEdges() [][3]interface{} {
	edgeMap := make(map[string][3]interface{})
	
	for vertex, edges := range g.data {
		for _, edge := range edges {
			// 辺を正規化（小さい方の頂点を最初にする）
			v1, v2 := vertex, edge.vertex
			if v1 > v2 {
				v1, v2 = v2, v1
			}
			
			// 辺のキーを作成
			key := v1 + "_" + v2
			
			// 既に追加されていなければ追加
			if _, exists := edgeMap[key]; !exists {
				edgeMap[key] = [3]interface{}{v1, v2, edge.weight}
			}
		}
	}
	
	// マップからスライスに変換
	edges := make([][3]interface{}, 0, len(edgeMap))
	for _, edge := range edgeMap {
		edges = append(edges, edge)
	}
	
	return edges
}

// GetNeighbors は指定された頂点の隣接ノードと辺の重みのスライスを返す
func (g *GraphData) GetNeighbors(vertex string) []Edge {
	if neighbors, exists := g.data[vertex]; exists {
		return neighbors
	}
	return nil
}

// GetEdgeWeight は指定された2つの頂点間の辺の重みを返す
func (g *GraphData) GetEdgeWeight(vertex1, vertex2 string) *int {
	if neighbors, exists := g.data[vertex1]; exists {
		for _, edge := range neighbors {
			if edge.vertex == vertex2 {
				return &edge.weight
			}
		}
	}
	return nil
}

// AddVertex は新しい頂点をグラフに追加する
func (g *GraphData) AddVertex(vertex string) bool {
	if _, exists := g.data[vertex]; !exists {
		g.data[vertex] = []Edge{}
	}
	return true
}

// AddEdge は両頂点間に辺を追加する
func (g *GraphData) AddEdge(vertex1, vertex2 string, weight int) bool {
	// 頂点が存在しなければ追加
	if _, exists := g.data[vertex1]; !exists {
		g.AddVertex(vertex1)
	}
	if _, exists := g.data[vertex2]; !exists {
		g.AddVertex(vertex2)
	}
	
	// vertex1 -> vertex2 の辺を追加/更新
	edgeExistsV1V2 := false
	for i, edge := range g.data[vertex1] {
		if edge.vertex == vertex2 {
			g.data[vertex1][i].weight = weight
			edgeExistsV1V2 = true
			break
		}
	}
	if !edgeExistsV1V2 {
		g.data[vertex1] = append(g.data[vertex1], Edge{vertex: vertex2, weight: weight})
	}
	
	// vertex2 -> vertex1 の辺を追加/更新
	edgeExistsV2V1 := false
	for i, edge := range g.data[vertex2] {
		if edge.vertex == vertex1 {
			g.data[vertex2][i].weight = weight
			edgeExistsV2V1 = true
			break
		}
	}
	if !edgeExistsV2V1 {
		g.data[vertex2] = append(g.data[vertex2], Edge{vertex: vertex1, weight: weight})
	}
	
	return true
}

// IsEmpty はグラフが空かどうかを返す
func (g *GraphData) IsEmpty() bool {
	return len(g.data) == 0
}

// Clear はグラフを空にする
func (g *GraphData) Clear() bool {
	g.data = make(map[string][]Edge)
	return true
}

// GetMST は最小全域木を求める
func (g *GraphData) GetMST(startVertex string) [][3]interface{} {
	vertices := g.GetVertices()
	if len(vertices) == 0 {
		return [][3]interface{}{} // グラフが空
	}
	
	if startVertex == "" && len(vertices) > 0 {
		startVertex = vertices[0]
	} else if _, exists := g.data[startVertex]; !exists {
		fmt.Printf("ERROR: 開始頂点 %s はグラフに存在しません。\n", startVertex)
		return nil
	}
	
	// MSTに含まれる頂点のセット
	inMST := make(map[string]bool)
	// 優先度付きキュー
	minHeap := &MinHeap{}
	heap.Init(minHeap)
	// MSTを構成する辺のリスト
	mstEdges := [][3]interface{}{}
	// 各頂点への最小コストとその遷移元の頂点を記録
	minCost := make(map[string]int)
	parent := make(map[string]string)
	
	// 初期化
	for _, v := range vertices {
		minCost[v] = int(^uint(0) >> 1) // max int
		parent[v] = ""
	}
	
	// 開始頂点の処理
	minCost[startVertex] = 0
	heap.Push(minHeap, &MinHeapItem{
		cost:     0,
		vertex:   startVertex,
		fromEdge: "",
		index:    0,
	})
	
	for minHeap.Len() > 0 {
		// 最小コストの辺を持つ頂点を取り出す
		item := heap.Pop(minHeap).(*MinHeapItem)
		currentVertex := item.vertex
		fromVertex := item.fromEdge
		
		// 既にMSTに含まれている頂点であればスキップ
		if inMST[currentVertex] {
			continue
		}
		
		// 現在の頂点をMSTに追加
		inMST[currentVertex] = true
		
		// MSTに追加された辺を記録 (開始頂点以外)
		if fromVertex != "" {
			// fromVertex から currentVertex への辺の重みを取得
			weight := g.GetEdgeWeight(fromVertex, currentVertex)
			if weight != nil {
				// 辺を正規化して追加
				v1, v2 := fromVertex, currentVertex
				if v1 > v2 {
					v1, v2 = v2, v1
				}
				mstEdges = append(mstEdges, [3]interface{}{v1, v2, *weight})
			}
		}
		
		// 現在の頂点に隣接する頂点を調べ、MSTへの追加コストを更新
		neighbors := g.GetNeighbors(currentVertex)
		if neighbors != nil {
			for _, edge := range neighbors {
				neighbor := edge.vertex
				weight := edge.weight
				
				// 隣接頂点がまだMSTに含まれておらず、現在のコストよりも小さい場合
				if !inMST[neighbor] && weight < minCost[neighbor] {
					minCost[neighbor] = weight
					parent[neighbor] = currentVertex
					heap.Push(minHeap, &MinHeapItem{
						cost:     weight,
						vertex:   neighbor,
						fromEdge: currentVertex,
					})
				}
			}
		}
	}
	
	return mstEdges
}

func main() {
	fmt.Println("Prims TEST -----> start")
	graphData := NewGraphData()
	
	graphData.Clear()
	inputList := [][3]interface{}{
		{"A", "B", 4},
		{"B", "C", 3},
		{"B", "D", 2},
		{"D", "A", 1},
		{"A", "C", 2},
		{"B", "D", 2},
	}
	for _, input := range inputList {
		graphData.AddEdge(input[0].(string), input[1].(string), input[2].(int))
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	outputMst := graphData.GetMST("")
	for _, edge := range outputMst {
		fmt.Printf("Edge: %v - %v, Weight: %v\n", edge[0], edge[1], edge[2])
	}
	totalWeight := 0
	for _, edge := range outputMst {
		totalWeight += edge[2].(int)
	}
	fmt.Printf("最小全域木の合計重み: %d\n", totalWeight)
	
	graphData.Clear()
	inputList = [][3]interface{}{
		{"A", "B", 4},
		{"C", "D", 4},
		{"E", "F", 1},
		{"F", "G", 1},
	}
	for _, input := range inputList {
		graphData.AddEdge(input[0].(string), input[1].(string), input[2].(int))
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	outputMst = graphData.GetMST("")
	for _, edge := range outputMst {
		fmt.Printf("Edge: %v - %v, Weight: %v\n", edge[0], edge[1], edge[2])
	}
	totalWeight = 0
	for _, edge := range outputMst {
		totalWeight += edge[2].(int)
	}
	fmt.Printf("最小全域木の合計重み: %d\n", totalWeight)
	
	graphData.Clear()
	inputList = [][3]interface{}{
		{"A", "B", 4},
		{"B", "C", 3},
		{"D", "E", 5},
	}
	for _, input := range inputList {
		graphData.AddEdge(input[0].(string), input[1].(string), input[2].(int))
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	outputMst = graphData.GetMST("")
	for _, edge := range outputMst {
		fmt.Printf("Edge: %v - %v, Weight: %v\n", edge[0], edge[1], edge[2])
	}
	totalWeight = 0
	for _, edge := range outputMst {
		totalWeight += edge[2].(int)
	}
	fmt.Printf("最小全域木の合計重み: %d\n", totalWeight)
	
	graphData.Clear()
	inputList = [][3]interface{}{}
	for _, input := range inputList {
		graphData.AddEdge(input[0].(string), input[1].(string), input[2].(int))
	}
	fmt.Println("\nグラフの頂点:", graphData.GetVertices())
	fmt.Println("グラフの辺 (重み付き):", graphData.GetEdges())
	outputMst = graphData.GetMST("")
	for _, edge := range outputMst {
		fmt.Printf("Edge: %v - %v, Weight: %v\n", edge[0], edge[1], edge[2])
	}
	totalWeight = 0
	for _, edge := range outputMst {
		totalWeight += edge[2].(int)
	}
	fmt.Printf("最小全域木の合計重み: %d\n", totalWeight)
	
	fmt.Println("\nPrims TEST <----- end")
}
